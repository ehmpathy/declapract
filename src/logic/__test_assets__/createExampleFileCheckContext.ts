import { FileCheckContext } from '../../domain/objects/FileCheckContext';

export const createExampleFileCheckContext = () =>
  new FileCheckContext({
    relativeFilePath: 'some/file/path.ts',
    projectVariables: {},
    declaredFileContents: '__DECLARED_CONTENTS__',
    required: true,
    getProjectRootDirectory: () => '/some/project/root',
  });
