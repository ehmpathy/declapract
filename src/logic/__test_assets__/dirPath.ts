export const testAssetsDirectoryPath = __dirname;
