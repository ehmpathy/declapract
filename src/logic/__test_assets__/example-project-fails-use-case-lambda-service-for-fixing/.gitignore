# all contents of dir are ignored since tests provision and fix the contents in here
*
!.gitignore
!declapract.use.yml

