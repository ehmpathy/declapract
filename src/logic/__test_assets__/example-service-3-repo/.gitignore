.declapract
node_modules
package-lock.json
