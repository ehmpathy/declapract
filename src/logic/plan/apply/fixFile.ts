import { FileCheckEvaluation } from '../../../domain';
import { readFileIfExistsAsync } from '../../../utils/fileio/readFileIfExistsAsync';
import { removeFileAsync } from '../../../utils/fileio/removeFileAsync';
import { writeFileAsync } from '../../../utils/fileio/writeFileAsync';
import { UnexpectedCodePathError } from '../../UnexpectedCodePathError';

/**
 * fix a file using the fix function of a check, based on an evaluation of the file
 *
 * input is a "checkEvaluation", because:
 * - it tells us this file _was_ checked before this was called
 * - it has the exact "path" that the check was evaluated on (instead of just a glob, like on the path)
 * - it has the reference to the declared check, which has the fix function
 */
export const fixFile = async ({
  evaluation,
  projectRootDirectory,
}: {
  evaluation: FileCheckEvaluation;
  projectRootDirectory: string;
}) => {
  if (!evaluation.fix) throw new UnexpectedCodePathError('fixFile called on an eval that doesnt have a fix defined');
  const filePath = `${projectRootDirectory}/${evaluation.path}`;
  const fileContents = await readFileIfExistsAsync({ filePath });
  const desiredContents = await evaluation.fix(fileContents);
  desiredContents
    ? await writeFileAsync({ path: filePath, content: desiredContents })
    : await removeFileAsync({ path: filePath });
};
