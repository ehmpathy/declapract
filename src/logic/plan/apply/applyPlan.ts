import chalk from 'chalk';
import indentString from 'indent-string';
import { withNot } from 'simple-type-guards';
import { FileActionPlan, hasFailed, isFixableCheck, RequiredAction } from '../../../domain';
import { UnexpectedCodePathError } from '../../UnexpectedCodePathError';
import { getColoredPlanTitle } from '../display/color/getColoredPlanTitle';
import { sortFileCheckEvaluationsByPracticeRef } from '../sortFileCheckEvaluationsByPracticeRef';
import { sortFilePracticeEvaluationsByPracticeName } from '../sortFilePracticeEvaluationsByPracticeName';
import { fixFile } from './fixFile';

/**
 * for each "fixable" and "failed" check in the plan, apply the fix
 */
export const applyPlan = async ({
  plan,
  projectRootDirectory,
}: {
  plan: FileActionPlan;
  projectRootDirectory: string;
}) => {
  // sanity check that the plan has fixable actions
  if (plan.action !== RequiredAction.FIX_AUTOMATIC)
    throw new UnexpectedCodePathError('asked to apply a plan which does not have action FIX_AUTOMATIC specified', {
      plan,
    }); // should have been filtered out by now

  // print out the title for this file
  const title = getColoredPlanTitle({ plan });
  console.log(`  * ${title}`); // tslint:disable-line: no-console

  // for each failed evaluation check, output that we're applying it
  for (const evaluation of plan.evaluations.sort(sortFilePracticeEvaluationsByPracticeName)) {
    // grab the failed, fixable checks
    const failedFixableChecks = evaluation.checks
      .filter(hasFailed)
      .filter(isFixableCheck)
      .sort(sortFileCheckEvaluationsByPracticeRef);

    // apply each of them
    await Promise.all(
      failedFixableChecks.map(async (evaluation) => {
        await fixFile({ evaluation, projectRootDirectory });
        const statusToken = chalk.green('âœ“');
        const fixabilityToken = chalk.gray('(fix:applied)');
        console.log(indentString(`${statusToken} practice:${evaluation.practiceRef} ${fixabilityToken}`, 4)); // tslint:disable-line: no-console
      }),
    );
  }
};
