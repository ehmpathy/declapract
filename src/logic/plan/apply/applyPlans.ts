import { FileActionPlan } from '../../../domain';
import { applyPlan } from './applyPlan';

export const applyPlans = async ({
  plans,
  projectRootDirectory,
}: {
  plans: FileActionPlan[];
  projectRootDirectory: string;
}) => {
  const sortedPlans = plans.sort((a, b) => (a.path < b.path ? -1 : 1));
  for (const plan of sortedPlans) {
    await applyPlan({ plan, projectRootDirectory }); // one at a time, sequentially
  }
};
