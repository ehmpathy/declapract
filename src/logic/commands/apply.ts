import { RequiredAction } from '../../domain';
import { UnexpectedCodePathError } from '../UnexpectedCodePathError';
import { applyPlans } from '../usage/plan/apply/applyPlans';
import { filterPracticeEvaluationsFromPlans } from '../usage/plan/filterPracticeEvaluationsFromPlans';
import { getPlansForProject } from '../usage/plan/getPlansForProject';
import { readUsePracticesConfig } from '../usage/readUsePracticesConfig';

export const apply = async ({
  usePracticesConfigPath,
  filter,
}: {
  usePracticesConfigPath: string;
  filter?: {
    practiceNames?: string[];
  };
}) => {
  // read the usage config
  const config = await readUsePracticesConfig({ configPath: usePracticesConfigPath });

  // grab the selected use case's practices
  const useCase = config.declared.useCases.find((useCase) => useCase.name === config.useCase);
  if (!useCase)
    throw new UnexpectedCodePathError(
      'requested use case was not defined on config. should have thrown an error when processing the config by now',
    );

  // get plans for this project
  const plans = await getPlansForProject({
    practices: useCase.practices,
    projectRootDirectory: config.rootDir,
    projectVariables: config.variables,
  });

  // filter out the practices to the ones that can be automatically applied + for the practices specified
  const plansToApply = (
    await filterPracticeEvaluationsFromPlans({
      plans,
      filter: { byFixable: true, byPracticeNames: filter?.practiceNames ?? undefined },
    })
  ).filter((plan) => plan.action === RequiredAction.FIX_AUTOMATIC);

  // display the plans
  await applyPlans({ plans: plansToApply, projectRootDirectory: config.rootDir, projectVariables: config.variables });
};
