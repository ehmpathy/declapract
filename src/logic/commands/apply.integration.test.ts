import { removeFileAsync } from '../../utils/fileio/removeFileAsync';
import { listFilesInDirectory } from '../../utils/filepaths/listFilesInDirectory';
import { testAssetsDirectoryPath } from '../__test_assets__/dirPath';
import { apply } from './apply';

const logSpy = jest.spyOn(console, 'log');

describe('apply', () => {
  beforeEach(() => jest.clearAllMocks());
  it('should be able to apply for an example project', async () => {
    const targetDir = `${testAssetsDirectoryPath}/example-project-fails-use-case-lambda-service-for-fixing`;

    // delete the generated files to get the testing environment to the expected state
    const expectedFilesInDirectory = ['.gitignore', 'declapract.use.yml'];
    const filesInDirectory = await listFilesInDirectory(targetDir);
    await Promise.all(
      filesInDirectory
        .filter((path) => !expectedFilesInDirectory.includes(path))
        .map((path) => removeFileAsync({ path: `${targetDir}/${path}` })),
    );
    const filesInDirectoryNow = await listFilesInDirectory(targetDir);
    expect(filesInDirectoryNow).toEqual(expectedFilesInDirectory); // fail if the test environment is not correct

    // now apply
    await apply({
      usePracticesConfigPath: `${targetDir}/declapract.use.yml`,
    });
    expect(logSpy.mock.calls).toMatchSnapshot();
  });
});
