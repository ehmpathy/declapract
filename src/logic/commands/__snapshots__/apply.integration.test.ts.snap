// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`apply should be able to apply for an example project 1`] = `
Array [
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m .prettierignore[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:prettier.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m .terraform-version[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m prettier.config.js[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:prettier.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m provision/terraform/environments/dev/main.tf[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m provision/terraform/environments/dev/versions.tf[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m provision/terraform/environments/prod/main.tf[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m provision/terraform/environments/prod/versions.tf[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m provision/terraform/product/variables.tf[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:terraform.best-practice [90m(fix:applied)[39m",
  ],
  Array [
    "  * [1m[1m[33m[FIX_AUTOMATIC][39m[1m serverless.yml[22m",
  ],
  Array [
    "    [32mâœ“[39m practice:serverless.best-practice [90m(fix:applied)[39m",
  ],
]
`;
